### PR TITLE
Add nilight support proposal

### PR DESCRIPTION
**Pull request by mp3butcher**
fix proposal for https://gitlab.com/OpenMW/openmw/issues/2391  and 
perhaps https://gitlab.com/OpenMW/openmw/issues/3011

Don't know how to acces esm::record for those lights
NiDirectionalLight are just classic NiLight so there's no radius, so i put radius to quite arbitrary values. 
For these lights attenuation parameter are setted at instanciation
Other cases have attenuation parameters setted so nothing special to do

Problem:
- SceneUtil::LightSource radius is hardcoded to some arbitrary values(1000 for light attenuation compute 10000 radius for lighsource radius)
- no light controller implemented
https://streamable.com/qs2b4
Capture d’écran de 2019-08-11 02-24-29
Capture d’écran de 2019-08-07 02-15-53

### CODE REVIEW

**Comment by mp3butcher** [comment]
I would require more mods with nilights for testing...

File: components/nifosg/nifloader.cpp [hunk 1]
```diff
@@ -364,6 +366,67 @@ namespace NifOsg
...
+                const Nif::NiLight* nilight = static_cast<const Nif::NiLight*>(nifNode);
+                //TOFIX
+                float radius = std::max(1000.0f, nilight->boundXYZ.length());
+                //store osg::LightSource and convert it at instanciation to SceneUtil::LightSource
+                osg::ref_ptr<osg::LightSource> lightSource (new osg::LightSource);
+                //record this arbitrary radius for instanciation setup
+                lightSource->setUserData(new osg::FloatValueObject("radius",radius));
+                osg::ref_ptr<osg::Light> light (new osg::Light);
+
+                const Nif::NiPointLight* nipointlight = dynamic_cast<const Nif::NiPointLight*>(nifNode);
```
Comment (Capostrophic) [inline components/nifosg/nifloader.cpp]: Bad. Very bad. Terrible. We don't need dynamic casts to know the type of a NIF record. Just add RC_NiPointLight, RC_NiSpotLight etc. record types, update makeFactory accordingly and use them here.

Comment (mp3butcher) [reply components/nifosg/nifloader.cpp]: AFAIK Nif::NiPointLight has recType==RC_Nilight so the dynamicast is mandatory

Comment (Capostrophic) [reply components/nifosg/nifloader.cpp]: Please reread what I said. No, it's by no means mandatory.

Comment (mp3butcher) [reply components/nifosg/nifloader.cpp]: ...so tell me what you'd put...

Comment (mp3butcher) [reply components/nifosg/nifloader.cpp]: ok i'll add others RC_types

Comment (akortunov) [reply components/nifosg/nifloader.cpp]: Just modify entries in the niffile.cpp, so different records will use different types:

```
    newFactory.insert(makeEntry("NiAmbientLight",             &construct <NiLight>                     , RC_NiAmbientLight                       ));
    newFactory.insert(makeEntry("NiDirectionalLight",         &construct <NiLight>                     , RC_NiDirectionalLight                       ));
    newFactory.insert(makeEntry("NiPointLight",               &construct <NiPointLight>                , RC_NiPointLight                       ));
    newFactory.insert(makeEntry("NiSpotLight",                &construct <NiSpotLight>                 , RC_NiSpotLight                       ));```

Comment (mp3butcher) [reply components/nifosg/nifloader.cpp]: done

**Comment by akortunov** [comment]
About controllers:
1. NiLightDimmerController - a float value interpolation controller, should be very similar to NiAlphaController and NiVisController.
Allows to modify dimmer value depenting on simulation time (so ambient, diffuse and specular components will be changed as well).
2. NiLightIntensityController - a float value interpolation controller, should be very similar to NiAlphaController and NiVisController.
Supposedly allows to modify light radius or brightness (needs testing).
3. NiLightColorController - a vector interpolation controller, should be very similar to NiMaterialColorController.
Allows to modify a light color depending on similation time.

**Comment by psi29a** [comment]
@Capostrophic remarks?

@mp3butcher you still around to carry on? Looks promising.

**Comment by akortunov** [comment]
Known issues:
1. Light controllers are still missing.
2. We still do not know how to calculate light radius properly.
3. An approach with visitor is questionable.
4. Code itself requires a cleanup.

**Comment by psi29a** [comment]
Right, so unless those are taken care of, I'm going to close this for now. @mp3butcher feel free to reopen if you wish to work further on this.

**Comment by mp3butcher** [comment]
> Known issues:
> 
>     1. Light controllers are still missing.
For further pr ... don't have samples to test...
> 
>     2. We still do not know how to calculate light radius properly.
For further pr, a constant is best guess we could

> 
>     3. An approach with visitor is questionable.
In what we please? We haven't discuss that

> 
>     4. Code itself requires a cleanup.
What clean up? review please

**Pull request closed without merging**